### ISSUE TITLE
GPG-signed commits

### ISSUE THREAD

**Issue by parisk** [libgit2/pygit2#647]
Is there any way to sign commits (and tags maybe) with pygit2? I performed a quick search about `gpg` and `signed` both in the repository and documentation website and could not find any related document.

**Comment by parisk** [comment]
Sorry for nagging here. Can any of the maintainers share if this can happen right now, if there is any kind of workaround for this or, there are any features that need to be implemented first?

Thanks 🙂 .

**Comment by webknjaz** [comment]
I assume that it may be possible by manually assigning [`Commit.gpg_signature`](https://www.pygit2.org/objects.html#Commit.gpg_signature).

cc @jdavid

**Comment by webknjaz** [comment]
Looks like I was wrong. pygit2 only exposes already existing signatures.

But I've collected some pointers:
* It seems like more code need to be added around https://github.com/libgit2/pygit2/blob/0462c5ca1c9147804e15d594aafc086d7e4e3f98/src/repository.c#L1001 to optionally sign commits
* Producing a signed commit will probably involve invoking [`git_commit_create_buffer`](https://libgit2.org/libgit2/#HEAD/group/commit/git_commit_create_buffer) to create an in-memory commit in a [`git_buf`](https://libgit2.org/libgit2/#HEAD/type/git_buf) object having unsigned `commit_content` that will be passed to [`git_commit_create_with_signature`](https://libgit2.org/libgit2/#HEAD/group/commit/git_commit_create_with_signature). After this is done, I assume we'll need to update the ref manually because unlike `git_commit_create`, `git_commit_create_with_signature` does not have `update_ref`.
* Also we need to set a signing callback in [`git_rebase_options`](https://libgit2.org/libgit2/#HEAD/type/git_rebase_options): https://github.com/libgit2/libgit2/blob/b0692d6b3e818b9389295d7d33a0601143cc0c16/include/git2/rebase.h#L84 / [`git_commit_signing_cb`](https://libgit2.org/libgit2/#HEAD/group/callback/git_commit_signing_cb)

I'd expect the public API of pygit2 to just accept some args via [`Repository.create_commit()`](https://www.pygit2.org/objects.html#pygit2.Repository.create_commit) toggling the signing as well as maybe some global settings/callbacks for defaults.

**Comment by kuwv** [comment]
I really think is a really important feature and in my opinion should be prioritized. Without it it's too easy to impersonate people even on GitHub. Without this capability users are forced between choosing security or automation but not both.

**Comment by kuwv** [comment]
https://github.com/libgit2/libgit2/pull/3673

**Comment by kuwv** [comment]
https://github.com/libgit2/pygit2/pull/1142

Example:
```
""Prototype GnuPG signing."""

import logging
import os

from gnupg import GPG
from pygit2 import Repository, discover_repository

log = logging.getLogger(__name__)
log.setLevel(logging.WARNING)
log.addHandler(logging.StreamHandler())

CURRENT_DIR = os.getcwd()
REPO_DIR = discover_repository(CURRENT_DIR)
repo = Repository(REPO_DIR)

with open('/dev/stdout') as fd:
    tty_path = os.ttyname(fd.fileno())
    os.environ['GPG_TTY'] = tty_path

GNUPGHOME = os.path.join(os.path.expanduser('~'), '.gnupg-test')
if not os.path.exists(GNUPGHOME):
    log.info('making gnupghome')
    os.mkdir(GNUPGHOME)
    os.chmod(GNUPGHOME, 0o700)
    gpg = GPG(gnupghome=GNUPGHOME)

    log.info('generating a gpg key')
    input_data = gpg.gen_key_input(
        name_real='Jesse P. Johnson',
        name_email='jesse@example.com',
        name_comment='This is a test.',
        key_type='RSA',
        key_length=2048,
        subkey_type='RSA',
        subkey_length=2048,
        passphrase='secret',
    )
    key = gpg.gen_key(input_data)
    log.info(type(key), vars(key))
else:
    gpg = GPG(gnupghome=GNUPGHOME)
    log.info('found gpg key')

if __name__ == '__main__':
    # print('get reference to current branch')
    # ref = repo.head.name

    log.info('prepare commit info')
    message = 'ci: generate signed commit'

    log.info('populate index')
    author = repo.default_signature
    commiter = repo.default_signature
    repo.index.add(os.path.basename(__file__))
    repo.index.write()
    tree = repo.index.write_tree()
    parents = [repo.head.target]
    encoding = 'utf-8'

    log.info('create commit object')
    commit_string = repo.create_commit_string(
        author,
        commiter,
        message,
        tree,
        parents,
        encoding,
    )

    log.info('sign commit')
    signed_commit = gpg.sign(
        commit_string,
        passphrase='secret',
        detach=True,
    )
    # XXX: you need a detatched signature but verification fails
    # verified = gpg.verify(signed_commit.data)
    # print('Verified' if verified else 'Unverified')

    commit = repo.create_commit_with_signature(
        commit_string, signed_commit.data.decode('utf-8')
    )

    # TODO: sure there is a better way to pass the message
    obj = repo.get(commit)
    repo.head.set_target(commit, obj.message)
```

**Comment by vsajip** [comment]
Hmmm, can't see anything obviously wrong. It might be worth turning on logging for the `gnupg` logger and seeing what that shows. Do you get the same result if you do `gpg.verify_file(io.BytesIO(signed_commit.data))`? The test `test_signature_verification` in the source distribution's `test_gnupg.py` tests verifying a detached signature, and that test seems to pass. :thinking:

**Comment by vsajip** [comment]
It shouldn't matter, but ...

**Comment by kuwv** [comment]
> It shouldn't matter, but ... wait, I'm talking about using verify(), not sign().

Yeah, I meant the tandems of `sign`/`verify` vs `sign_file`/`verify_file`

> Thought about it some more ... when verifying a detached signature, two things are needed: the signature data and the signed data. The signature is in `signed_commit.data`, but the data you signed is in `commit_string`. So, I suggest writing that to a temporary file, then passing the name of that into `gpg.verify(signed_commit.data, data_filename=path_to_temp_file)`.

This worked, but I had to do it using the builtin `_make_binary_stream`. I wonder if my encodings are not using `utf-8` somewhere.

Anyway, this was really helpful @vsajip. Thank you so much!

**Comment by kuwv** [comment]
@jdavid This issue can be closed now.

https://github.com/libgit2/libgit2/issues/6293

**Comment by chet-manley** [comment]
```
# XXX: this attempt did not work
# verified = gpg.verify_file(BytesIO(signature.data), f.name)
```
After writing, the file pointer is at the end of the file, and there is nothing left to read.
To get a proper verification, open the temporary file with mode `w+b` and `file.seek(0)` after writing out the file.

**Comment by kuwv** [comment]
> ```
> # XXX: this attempt did not work
> # verified = gpg.verify_file(BytesIO(signature.data), f.name)
> ```
> 
> After writing, the file pointer is at the end of the file, and there is nothing left to read. To get a proper verification, open the temporary file with mode `w+b` and `file.seek(0)` after writing out the file.

Yup, it works. Don't know why that didn't occur to me.
